src/wallet/wallet.cpp

```diff
@@ -426,27 +442,243 @@ ZcashdUnifiedSpendingKey CWallet::GenerateNewUnifiedSpendingKey() {
+            auto receivers = metadata->second.GetReceivers(j);
+            if (receivers.has_value()) {
+                if (receivers.value() == receiverTypes) {
+                    return std::make_pair(ufvk.value().Address(j, receiverTypes).value(), j);
+                } else {
+                    return AddressGenerationError::ExistingAddressMismatch;
+                }
+            }
+        }
+
+        // Find a working diversifier and construct the associated address.
+        auto foundAddress = ufvk.value().FindAddress(j, receiverTypes);
+        auto diversifierIndex = foundAddress.second;
+
+        // Persist the newly created address to the keystore
+        mapUfvkAddressMetadata[ufvkid].SetReceivers(diversifierIndex, receiverTypes);
+        CCryptoKeyStore::AddUnifiedAddress(ufvkid, diversifierIndex, foundAddress.first);
+
+        // Save the metadata for the generated address so that we can re-derive
+        // it in the future.
+        ZcashdUnifiedAddressMetadata addrmeta(ufvkid, diversifierIndex, receiverTypes);
+        if (fFileBacked && !CWalletDB(strWalletFile).WriteUnifiedAddressMetadata(addrmeta)) {
+            throw std::runtime_error(
+                    "CWallet::AdddUnifiedAddress(): Writing unified address metadata failed");
+        }
+
+        if (hasTransparent) {
+            // Regenerate the secret key for the transparent address and add it to
+            // the wallet.
+            auto seed = GetMnemonicSeed().value();
+            auto b44 = libzcash::Bip44AccountChains::ForAccount(seed, BIP44CoinType(), accountId).value();
+            auto key = b44.DeriveExternal(diversifierIndex.ToTransparentChildIndex().value()).value();
+            AddTransparentSecretKey(seed.Fingerprint(), key);
+        }
+
+        return foundAddress;
+    } else {
+        return AddressGenerationError::NoSuchAccount;
+    }
+}
+
+// NOTE: There is an important relationship between `LoadUnifiedFullViewingKey`
+// and `LoadUnifiedAddressMetadata`. In all cases, by the end of the
+// wallet-loading process, `mapUfvkAddressMetadata` needs to be fully populated
+// for each unified full viewing key. However, we cannot guarantee in what
+// order UFVKs and unified address and account metadata are loaded from disk.
+// Therefore:
+// * When we load a unified full viewing key from disk, we repopulate
+//   the cache of generated addresses by regenerating the address for each
+//   (diversifierIndex, {receiverType...}) pair in `mapUfvkAddressMetadata`
+//   that we have loaded so far.
+// * When we load a unified address metadata record from disk:
+//   - if we have not yet loaded the unified full viewing key, simply add
+//     an entry `mapUfvkAddressMetadata` so that we can restore the address
+//     once the UFVK has been loaded.
+//   - if we have already loaded the unified full viewing key from disk,
+//     regenerate the address from its metadata and add it to the keystore
```

Comment:
This seems horribly complicated, is there no better way?

Reply:
I added a rationale to the comment.